sdk/identity/Azure.Identity/src/ClientAssertionCredential.cs

```diff
@@ -0,0 +1,71 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Collections.Generic;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Azure.Core;
+using Azure.Core.Pipeline;
+using Microsoft.Identity.Client;
+
+namespace Azure.Identity
+{
+    internal class ClientAssertionCredential : TokenCredential
+    {
+        internal string TenantId { get; }
+        internal string ClientId { get; }
+        internal MsalConfidentialClient Client { get; }
+        internal bool AllowMultiTenantAuthentication { get; }
+
+        public ClientAssertionCredential(string tenantId, string clientId, Func<Task<string>> getAssertionCallback, ClientAssertionCredentialOptions options = default) :
+            this(tenantId, clientId, () => getAssertionCallback().GetAwaiter().GetResult(), options)
+        {
+        }
+
+        public ClientAssertionCredential(string tenantId, string clientId, Func<string> getAssertionCallback, ClientAssertionCredentialOptions options = default)
+        {
+            TenantId = tenantId;
+            ClientId = clientId;
+            AllowMultiTenantAuthentication = options?.AllowMultiTenantAuthentication ?? false;
+            Client = options?.MsalClient ?? new MsalConfidentialClient(CredentialPipeline.GetInstance(options), tenantId, clientId, getAssertionCallback, null, null, options?.IsLoggingPIIEnabled ?? false);
+        }
+
+        public override AccessToken GetToken(TokenRequestContext requestContext, CancellationToken cancellationToken)
+        {
+            using CredentialDiagnosticScope scope = Client.Pipeline.StartGetTokenScope("ClientAssertionCredential.GetToken", requestContext);
+
+            try
+            {
+                var tenantId = TenantIdResolver.Resolve(TenantId, requestContext, AllowMultiTenantAuthentication);
+
+                AuthenticationResult result = Client.AcquireTokenForClientAsync(requestContext.Scopes, tenantId, false, cancellationToken).EnsureCompleted();
```

Comment:
For later - maybe we should consider encapsulating the TenantIdResolver inside the `MsalConfidentialClient` and `MsalPublicClient` since it's become a duplicate piece of code everywhere.

Reply:
Created issue https://github.com/Azure/azure-sdk-for-net/issues/23866 to address this later.